examples/Cassie/osc_jump/pelvis_orientation_traj_generator.cc

```diff
@@ -0,0 +1,69 @@
+#include "pelvis_orientation_traj_generator.h"
+
+#include "multibody/multibody_utils.h"
+
+using std::cout;
+using std::endl;
+
+using Eigen::Map;
+using Eigen::MatrixXd;
+using Eigen::Vector2d;
+using Eigen::Vector3d;
+using Eigen::VectorXd;
+using std::string;
+
+using dairlib::multibody::createContext;
+using dairlib::systems::OutputVector;
+using drake::multibody::BodyFrame;
+using drake::multibody::Frame;
+using drake::multibody::MultibodyPlant;
+using drake::systems::BasicVector;
+using drake::systems::Context;
+using drake::systems::DiscreteUpdateEvent;
+using drake::systems::DiscreteValues;
+using drake::systems::EventStatus;
+using drake::trajectories::ExponentialPlusPiecewisePolynomial;
+using drake::trajectories::PiecewisePolynomial;
+using drake::trajectories::Trajectory;
+
+namespace dairlib::examples::Cassie::osc_jump {
+
+PelvisOrientationTrajGenerator::PelvisOrientationTrajGenerator(
+    const drake::multibody::MultibodyPlant<double>& plant,
+    const drake::trajectories::PiecewisePolynomial<double>& orientation_traj,
+    std::string traj_name, double time_offset)
+    : plant_(plant), traj_(orientation_traj) {
+  PiecewisePolynomial<double> empty_pp_traj(VectorXd(0));
+  Trajectory<double>& traj_inst = empty_pp_traj;
+
+  this->set_name("traj_name");
+  this->DeclareAbstractOutputPort("traj_name", traj_inst,
+                                  &PelvisOrientationTrajGenerator::CalcTraj);
+
+  // Input/Output Setup
+  state_port_ =
+      this->DeclareVectorInputPort(OutputVector<double>(plant_.num_positions(),
+                                                        plant_.num_velocities(),
+                                                        plant_.num_actuators()))
+          .get_index();
+
+  // Shift trajectory by time_offset
+  traj_.shiftRight(time_offset);
+}
+
+void PelvisOrientationTrajGenerator::CalcTraj(
+    const drake::systems::Context<double>& context,
+    drake::trajectories::Trajectory<double>* traj) const {
+  // Read in current state
+  const OutputVector<double>* robot_output =
+      (OutputVector<double>*)this->EvalVectorInput(context, state_port_);
+  double timestamp = robot_output->get_timestamp();
+
+  auto* casted_traj =
+      (PiecewisePolynomial<double>*)dynamic_cast<PiecewisePolynomial<double>*>(
+          traj);
+
+  *casted_traj = traj_.slice(traj_.get_segment_index(timestamp), 1);
```

Comment:
Reading this, it's not totally clear to me what it's doing. Is it extracting a subtrajectory for the 2 knot points around the current time? I'm guessing that the reason is to avoid pushing the entire trajectory out through the output port. If so, is 1 segment guaranteed to be enough?

Reply:
I think this was a remnant of me trying to speed up the controller - didn't end up making a big difference. I'll change it to return the entire trajectory to make it clearer but I think either method should work. The duration of single segment should definitely be longer than a control loop and drake trajectories are "extended" past the duration bounds.